br/pkg/lightning/importer/dup_detect.go

```diff
@@ -0,0 +1,387 @@
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/extsort"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+type dupDetector struct {
+	tr     *TableImporter
+	rc     *Controller
+	cp     *checkpoints.TableCheckpoint
+	logger log.Logger
+}
+
+func (d *dupDetector) run(
+	ctx context.Context,
+	workingDir string,
+	ignoreRows extsort.ExternalSorter,
+) (retErr error) {
+	var numDups int64
+	task := d.logger.Begin(zap.InfoLevel, "duplicate detection")
+	defer func() {
+		task.End(zap.ErrorLevel, retErr, zap.Int64("numDups", numDups))
+	}()
+
+	sorter, err := extsort.OpenDiskSorter(workingDir, &extsort.DiskSorterOptions{
+		Concurrency: d.rc.cfg.App.RegionConcurrency,
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer func() {
+		_ = sorter.CloseAndCleanup()
+	}()
+
+	detector := duplicate.NewDetector(sorter, d.logger)
+	if err := d.addKeys(ctx, detector); err != nil {
+		return errors.Trace(err)
+	}
+
+	handlerConstructor := makeDupHandlerConstructor(ignoreRows, d.rc.cfg.TikvImporter.OnDuplicate)
+	numDups, err = detector.Detect(ctx, &duplicate.DetectOptions{
+		Concurrency:        d.rc.cfg.App.RegionConcurrency,
+		HandlerConstructor: handlerConstructor,
+	})
+	return errors.Trace(err)
+}
+
+func makeDupHandlerConstructor(
+	sorter extsort.ExternalSorter, onDup string,
+) duplicate.HandlerConstructor {
+	switch onDup {
+	case config.ErrorOnDup:
+		return func(ctx context.Context) (duplicate.Handler, error) {
+			return &errorOnDup{}, nil
+		}
+	case config.ReplaceOnDup:
+		return func(ctx context.Context) (duplicate.Handler, error) {
+			w, err := sorter.NewWriter(ctx)
+			if err != nil {
+				return nil, err
+			}
+			return &replaceOnDup{w: w}, nil
+		}
+	case config.IgnoreOnDup:
+		return func(ctx context.Context) (duplicate.Handler, error) {
+			w, err := sorter.NewWriter(ctx)
+			if err != nil {
+				return nil, err
+			}
+			return &ignoreOnDup{w: w}, nil
+		}
+	default:
+		panic(fmt.Sprintf("unexpected on-duplicate strategy: %s", onDup))
+	}
+}
+
+var (
+	_ duplicate.Handler = &errorOnDup{}
+	_ duplicate.Handler = &replaceOnDup{}
+	_ duplicate.Handler = &ignoreOnDup{}
+)
+
+type errorOnDup struct{}
+
+func (errorOnDup) Begin(key []byte) error {
+	// TODO: add more useful information to the error message.
+	return errors.Errorf("duplicate key detected: %X", key)
+}
+
+func (errorOnDup) Append(_ []byte) error { return nil }
+func (errorOnDup) End() error            { return nil }
+func (errorOnDup) Close() error          { return nil }
+
+type replaceOnDup struct {
+	// All keyIDs except the last one will be written to w.
+	// keyID written to w will be ignored during importing.
+	w     extsort.Writer
+	keyID []byte
+}
+
+func (h *replaceOnDup) Begin(_ []byte) error {
+	h.keyID = h.keyID[:0]
+	return nil
+}
+
+func (h *replaceOnDup) Append(keyID []byte) error {
+	if len(h.keyID) > 0 {
+		if err := h.w.Put(h.keyID, nil); err != nil {
+			return err
+		}
+	}
+	h.keyID = append(h.keyID[:0], keyID...)
```

Comment:
I want to learn what's the benefit that we copy the content of []byte, rather than let `h.keyID` point to the underlying array of `keyID` 😄

Reply:
OK I guess compared with the normal allocate-on-write usage, when the read is less frequent than write, let reader copy the value can reduce allocation.

Reply:
The `keyID` may be modified by `Detector` after the method returns. Here we want to retain the `keyID`, so we must copy it.